File: packages/contracts/src/incident.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const incidentCategorySchema = z.enum([
  "SECURITY",
  "SUSPICIOUS_ACTIVITY",
  "DAMAGE",
  "NOISE",
  "TRAFFIC",
  "LOST_ITEM",
  "EMERGENCY",
  "OTHER",
]);
export const incidentStatusSchema = z.enum(["OPEN", "IN_REVIEW", "RESOLVED", "CLOSED"]);

export type IncidentCategory = z.infer<typeof incidentCategorySchema>;
export type IncidentStatus = z.infer<typeof incidentStatusSchema>;

export const createIncidentInputSchema = z
  .object({
    category: incidentCategorySchema,
    title: z.string().trim().min(3).max(200),
    description: z.string().trim().min(3).max(4000),
    location: z.string().trim().min(1).max(200).optional(),
    occurredAt: z.string().datetime({ offset: true }),
    peopleInvolved: z.string().trim().max(1000).optional(),
  })
  .strict();

export type CreateIncidentInput = z.infer<typeof createIncidentInputSchema>;

export const updateIncidentInputSchema = z
  .object({
    status: incidentStatusSchema.optional(),
    actionTaken: z.string().trim().max(2000).optional(),
  })
  .strict()
  .refine((input) => Object.keys(input).length > 0, {
    message: "Kirim setidaknya satu perubahan laporan.",
  });

export type UpdateIncidentInput = z.infer<typeof updateIncidentInputSchema>;

export const incidentListQuerySchema = z.object({
  status: incidentStatusSchema.optional(),
  limit: z.coerce.number().int().min(1).max(50).default(20),
});

export const incidentSummarySchema = z.object({
  id: z.string().uuid(),
  category: incidentCategorySchema,
  title: z.string().min(1),
  location: z.string().nullable(),
  occurredAt: z.string().datetime({ offset: true }),
  status: incidentStatusSchema,
  reporterName: z.string().min(1),
  createdAt: z.string().datetime({ offset: true }),
});

export const incidentDetailSchema = incidentSummarySchema.extend({
  description: z.string().min(1),
  peopleInvolved: z.string().nullable(),
  actionTaken: z.string().nullable(),
});

export type IncidentSummary = z.infer<typeof incidentSummarySchema>;
export type IncidentDetail = z.infer<typeof incidentDetailSchema>;

export const incidentListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(incidentSummarySchema) }),
);

export const incidentDetailResponseSchema = dataEnvelopeSchema(
  z.object({ incident: incidentDetailSchema }),
);

export const incidentMutationResponseSchema = dataEnvelopeSchema(
  z.object({ incident: incidentDetailSchema }),
);

export type IncidentListResponse = z.infer<typeof incidentListResponseSchema>;
export type IncidentDetailResponse = z.infer<typeof incidentDetailResponseSchema>;
export type IncidentMutationResponse = z.infer<typeof incidentMutationResponseSchema>;

File: apps/mobile/lib/features/incident/domain/incident.dart
enum IncidentCategory {
  security,
  suspiciousActivity,
  damage,
  noise,
  traffic,
  lostItem,
  emergency,
  other,
}

IncidentCategory _categoryFromApi(Object? value) {
  switch (value) {
    case 'SECURITY':
      return IncidentCategory.security;
    case 'SUSPICIOUS_ACTIVITY':
      return IncidentCategory.suspiciousActivity;
    case 'DAMAGE':
      return IncidentCategory.damage;
    case 'NOISE':
      return IncidentCategory.noise;
    case 'TRAFFIC':
      return IncidentCategory.traffic;
    case 'LOST_ITEM':
      return IncidentCategory.lostItem;
    case 'EMERGENCY':
      return IncidentCategory.emergency;
    default:
      return IncidentCategory.other;
  }
}

extension IncidentCategoryApi on IncidentCategory {
  String toApi() {
    switch (this) {
      case IncidentCategory.security:
        return 'SECURITY';
      case IncidentCategory.suspiciousActivity:
        return 'SUSPICIOUS_ACTIVITY';
      case IncidentCategory.damage:
        return 'DAMAGE';
      case IncidentCategory.noise:
        return 'NOISE';
      case IncidentCategory.traffic:
        return 'TRAFFIC';
      case IncidentCategory.lostItem:
        return 'LOST_ITEM';
      case IncidentCategory.emergency:
        return 'EMERGENCY';
      case IncidentCategory.other:
        return 'OTHER';
    }
  }

  String get label {
    switch (this) {
      case IncidentCategory.security:
        return 'Keamanan';
      case IncidentCategory.suspiciousActivity:
        return 'Aktivitas mencurigakan';
      case IncidentCategory.damage:
        return 'Kerusakan';
      case IncidentCategory.noise:
        return 'Kebisingan';
      case IncidentCategory.traffic:
        return 'Lalu lintas';
      case IncidentCategory.lostItem:
        return 'Barang hilang';
      case IncidentCategory.emergency:
        return 'Darurat';
      case IncidentCategory.other:
        return 'Lainnya';
    }
  }
}

enum IncidentStatus { open, inReview, resolved, closed }

IncidentStatus _statusFromApi(Object? value) {
  switch (value) {
    case 'IN_REVIEW':
      return IncidentStatus.inReview;
    case 'RESOLVED':
      return IncidentStatus.resolved;
    case 'CLOSED':
      return IncidentStatus.closed;
    default:
      return IncidentStatus.open;
  }
}

extension IncidentStatusApi on IncidentStatus {
  String toApi() {
    switch (this) {
      case IncidentStatus.open:
        return 'OPEN';
      case IncidentStatus.inReview:
        return 'IN_REVIEW';
      case IncidentStatus.resolved:
        return 'RESOLVED';
      case IncidentStatus.closed:
        return 'CLOSED';
    }
  }

  String get label {
    switch (this) {
      case IncidentStatus.open:
        return 'Terbuka';
      case IncidentStatus.inReview:
        return 'Ditinjau';
      case IncidentStatus.resolved:
        return 'Selesai';
      case IncidentStatus.closed:
        return 'Ditutup';
    }
  }
}

class IncidentSummary {
  const IncidentSummary({
    required this.id,
    required this.category,
    required this.title,
    required this.location,
    required this.occurredAt,
    required this.status,
    required this.reporterName,
    required this.createdAt,
  });

  final String id;
  final IncidentCategory category;
  final String title;
  final String? location;
  final DateTime occurredAt;
  final IncidentStatus status;
  final String reporterName;
  final DateTime createdAt;

  factory IncidentSummary.fromJson(Map<String, dynamic> json) {
    return IncidentSummary(
      id: json['id'] as String,
      category: _categoryFromApi(json['category']),
      title: json['title'] as String,
      location: json['location'] as String?,
      occurredAt: DateTime.parse(json['occurredAt'] as String),
      status: _statusFromApi(json['status']),
      reporterName: json['reporterName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

class IncidentDetail {
  const IncidentDetail({
    required this.id,
    required this.category,
    required this.title,
    required this.location,
    required this.occurredAt,
    required this.status,
    required this.reporterName,
    required this.createdAt,
    required this.description,
    required this.peopleInvolved,
    required this.actionTaken,
  });

  final String id;
  final IncidentCategory category;
  final String title;
  final String? location;
  final DateTime occurredAt;
  final IncidentStatus status;
  final String reporterName;
  final DateTime createdAt;
  final String description;
  final String? peopleInvolved;
  final String? actionTaken;

  factory IncidentDetail.fromJson(Map<String, dynamic> json) {
    return IncidentDetail(
      id: json['id'] as String,
      category: _categoryFromApi(json['category']),
      title: json['title'] as String,
      location: json['location'] as String?,
      occurredAt: DateTime.parse(json['occurredAt'] as String),
      status: _statusFromApi(json['status']),
      reporterName: json['reporterName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      description: json['description'] as String,
      peopleInvolved: json['peopleInvolved'] as String?,
      actionTaken: json['actionTaken'] as String?,
    );
  }
}
